core/src/main/java/ai/timefold/solver/core/impl/solver/ConsumerSupport.java

```diff
@@ -42,6 +44,27 @@ void consumeIntermediateBestSolution(Solution_ bestSolution, BooleanSupplier isE
         }
     }
 
+    // Called on the Solver thread.
+    void consumeInitializedSolution(Solution_ initializedSolution) {
+        if (initializedSolutionConsumer != null) {
+            // TODO - Do we need to sync it?
+            try {
+                // Wait for the previous consumption to complete.
+                activeConsumption.acquire();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new IllegalStateException("Interrupted when waiting for the final best solution consumption.");
+            }
+            try {
+                initializedSolutionConsumer.accept(initializedSolution);
+            } catch (Throwable throwable) {
+                exceptionHandler.accept(problemId, throwable);
+            } finally {
+                activeConsumption.release();
+            }
+        }
```

Comment:
Isn't this a bit over-complicated? Why all the locking? We're calling a simple listener.

I'm not versed in this particular aspect of the code; perhaps @rsynek could explain?

Reply:
I agree regarding the complexity. The initial goal was to bring this up for discussion because I don't think there is a need for a synchronous call.

I have simplified the logic and will now wait for comments from @rsynek.

Reply:
Generally speaking, if the consumption happens on the solver thread and it's a one-time event, I don't see any need of locking.

On the other hand, all the BSEs are being consumed on a consumer thread that is paired with the solver thread.

The reason is that consuming the solution is typically an IO operation that can delay the Solver. That being said, I leave it up to you to decide whether we keep such behavior (for a single solution, it might not be such a big deal, after all).

If you decide to move the consumption on the consumer thread, then locking might be necessary. Take a look at the final BSE consumption; the principle should be very similar.

Reply:
The Init event should in theory be always the first one, so the consumer will always grab it. However, in case of real-time planning, there might be a pending BSE waiting for consumption, that's why acquiring the `activeConsumption` lock might be necessary.

After the previous BSE is consumed, the Init event should have the right of way.

Reply:
Very useful information indeed. That makes sense, and I'll address that.